### ISSUE TITLE
Strangest bug ever. Docker behaving inconsistently?

### ISSUE THREAD

**Issue by pmjohann** [vercel/pkg#1313]
Hey everyone,

I've just come across one of my strangest bugs ever.
I use pkg during a container build, and the very same build succeeds on my laptop (Docker for Mac) yet fails on my server.

This is the error I get on the server:

```
Sending build context to Docker daemon   76.8kB
Step 1/9 : FROM alpine@sha256:e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a AS builder
 ---> 14119a10abf4
Step 2/9 : COPY . /app
 ---> Using cache
 ---> 6f5031611fbe
Step 3/9 : WORKDIR /app
 ---> Using cache
 ---> dfa5faa56d48
Step 4/9 : ENV NPM_VERSION=7.19.1
 ---> Using cache
 ---> 8251b20c1d62
Step 5/9 : ENV PKG_VERSION=v5.3.2
 ---> Using cache
 ---> 53e3a45ea02e
Step 6/9 : RUN apk add --no-cache npm &&     npm i -g npm@$NPM_VERSION &&     npm i -g pkg@$PKG_VERSION &&     npm install
 ---> Using cache
 ---> 6ea4c87569f5
Step 7/9 : RUN pkg     --targets node16-alpine-arm64     --output /dist/app     index.js
 ---> Running in f069808a7f72
> pkg@5.3.2
> Fetching base Node.js binaries to PKG_CACHE_PATH




internal/child_process.js:403
    throw errnoException(err, 'spawn');
    ^

Error: spawn Unknown system error -8
    at ChildProcess.spawn (internal/child_process.js:403:11)
    at spawn (child_process.js:667:9)
    at Object.spawnWithSignal [as spawn] (child_process.js:871:17)
    at fabricate (/usr/local/lib/node_modules/pkg/lib-es5/fabricator.js:55:41)
    at Object.fabricateTwice (/usr/local/lib/node_modules/pkg/lib-es5/fabricator.js:120:5)
    at MultiStream._queue (/usr/local/lib/node_modules/pkg/lib-es5/producer.js:266:49)
    at MultiStream._next (/usr/local/lib/node_modules/pkg/node_modules/multistream/index.js:86:12)
    at Meter.onEnd (/usr/local/lib/node_modules/pkg/node_modules/multistream/index.js:129:12)
    at Object.onceWrapper (events.js:519:28)
    at Meter.emit (events.js:400:28) {
  errno: -8,
  code: 'Unknown system error -8',
  syscall: 'spawn'
}
The command '/bin/sh -c pkg     --targets node16-alpine-arm64     --output /dist/app     index.js' returned a non-zero code: 1
```

---

I created a [sample repo](https://github.com/pmjohann/pkgdebug) to reproduce the issue, I'd be really curious what happens if you try to build it.
I used exact versions for reproducibility:

- the base image (alpine:3.14.2)
- pkg (5.3.2)
- npm (7.19.1)

### Simple script to reproduce:

```bash
git clone https://github.com/pmjohann/pkgdebug.git /tmp/pkgdebug
cd /tmp/pkgdebug
docker build -t pkgdebug .
```

### Script to clean up:

```bash
cd $HOME
docker image rm pkgdebug
rm -rf /tmp/pkgdebug
```

The base image's (alpine:3.14.2) SHA256 digest is the same both on the server and my laptop:

```e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a```

Please let me know your results, this drives me crazy!
Thank you in advance.

**Comment by pmjohann** [comment]
### Update: 
I specified the base image by SHA256 digest in the [Dockerfile](https://github.com/pmjohann/pkgdebug/blob/main/Dockerfile) now as well instead of a tag to ensure test correctness.

### Additional version info:

#### Server:

- Kernel: 4.15.0-156-generic
- Docker: 20.10.8 (both server & client)

#### Laptop:

- Kernel: 5.10.47-linuxkit (provided by Docker for Mac)
- Docker: 20.10.8 (both server & client)

**Comment by pmjohann** [comment]
Before anyone asks both platforms are x86_64 and NOT Apple Silicon (M1).

**Comment by dongnguyenvt** [comment]
I can reproduce the issue but I've got slightly different error code
```
events.js:377
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:156:25)
    at writeGeneric (internal/stream_base_commons.js:147:3)
    at Socket._writeGeneric (net.js:787:11)
    at Socket._write (net.js:799:8)
    at writeOrBuffer (internal/streams/writable.js:358:12)
    at Socket.Writable.write (internal/streams/writable.js:303:10)
    at fabricate (/usr/local/lib/node_modules/pkg/lib-es5/fabricator.js:111:17)
    at Object.fabricateTwice (/usr/local/lib/node_modules/pkg/lib-es5/fabricator.js:120:5)
    at MultiStream._queue (/usr/local/lib/node_modules/pkg/lib-es5/producer.js:266:49)
    at MultiStream._next (/usr/local/lib/node_modules/pkg/node_modules/multistream/index.js:86:12)
Emitted 'error' event on Socket instance at:
    at emitErrorNT (internal/streams/destroy.js:106:8)
    at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    at processTicksAndRejections (internal/process/task_queues.js:82:21) {
  errno: -32,
  code: 'EPIPE',
  syscall: 'write'
}
```

- full log gist: https://gist.github.com/dongnguyenvt/93bb044b210beae6c36a346ce6beeeeb

changing to build node base image: `node:16-alpine3.14` the error is gone, but have some warnings about ENOENT (no directory)
```
Sending build context to Docker daemon  75.26kB
Step 1/9 : FROM node:16-alpine3.14 AS builder
16-alpine3.14: Pulling from library/node
a0d0a0d46f8b: Already exists 
147d971d18b7: Pull complete 
aac7c3bf47ea: Pull complete 
69d53bb06ef6: Pull complete 
Digest: sha256:62f442b1f2be995cb7352df38a77f936cbedf02ca3213f0cf8f61facf74d0d68
Status: Downloaded newer image for node:16-alpine3.14
 ---> aae0878eed12
Step 2/9 : COPY . /app
 ---> bd3f5c665ec4
Step 3/9 : WORKDIR /app
 ---> Running in c437b5067cdd
Removing intermediate container c437b5067cdd
 ---> a30d1fcbb696
Step 4/9 : ENV NPM_VERSION=7.19.1
 ---> Running in 084a05935453
Removing intermediate container 084a05935453
 ---> 9c78a8771859
Step 5/9 : ENV PKG_VERSION=v5.3.2
 ---> Running in b0c2df2a16b7
Removing intermediate container b0c2df2a16b7
 ---> 138db564156c
Step 6/9 : RUN npm i -g npm@$NPM_VERSION &&     npm i -g pkg@$PKG_VERSION &&     npm install
 ---> Running in d994ce5dfd6c

removed 9 packages, changed 38 packages, and audited 257 packages in 12s

11 packages are looking for funding
  run `npm fund` for details

2 vulnerabilities (1 moderate, 1 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
npm notice 
npm notice New minor version of npm available! 7.21.1 -> 7.23.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v7.23.0>
npm notice Run `npm install -g npm@7.23.0` to update!
npm notice 

added 155 packages, and audited 156 packages in 6s

16 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

added 1 package, and audited 2 packages in 600ms

found 0 vulnerabilities
Removing intermediate container d994ce5dfd6c
 ---> 09ee0ada4597
Step 7/9 : RUN pkg     --targets node16-alpine-arm64     --output /dist/app     index.js
 ---> Running in a012d6e4b46a
> pkg@5.3.2
> Fetching base Node.js binaries to PKG_CACHE_PATH




> Warning Failed to make bytecode node16-arm64 for file /snapshot/app/index.js error (spawn /root/.pkg-cache/v3.2/fetched-v16.8.0-linuxstatic-arm64 ENOENT)
> Warning Failed to make bytecode node16-arm64 for file /snapshot/app/node_modules/lodash/lodash.js error (spawn /root/.pkg-cache/v3.2/fetched-v16.8.0-linuxstatic-arm64 ENOENT)
Removing intermediate container a012d6e4b46a
 ---> 444312560f86
Step 8/9 : FROM alpine@sha256:e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a
 ---> 14119a10abf4
Step 9/9 : COPY --from=builder /dist /dist
 ---> cbe5fa58dc7e
Successfully built cbe5fa58dc7e
```

So I think there could be some incompatible of the js bundle.

Regards.

**Comment by adamkdean** [comment]
I've been seeing this with a number of different issues (on and off) when building on server, but fine on my local machine.

```
linuxstatic-arm64: Syntax error: "(" unexpected

...

Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
```

```
/pkg-cache/v3.2/fetched-v14.17.6-linuxstatic-arm64: 1: /pkg-cache/v3.2/fetched-v14.17.6-linuxstatic-arm64: Syntax error: word unexpected (expecting ")")
/pkg-cache/v3.2/fetched-v14.17.6-linuxstatic-arm64: 1: /pkg-cache/v3.2/fetched-v14.17.6-linuxstatic-arm64: Syntax error: word unexpected (expecting ")")

...

[91mevents.js:292
      throw er; // Unhandled 'error' event
      ^

Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
```
